libgnucash/app-utils/options.scm

```diff
@@ -25,2106 +25,97 @@
-
-  (define (dispatch key)
-    (case key
-      ((lookup) lookup-option)
-      ((register-option) register-option)
-      ((unregister-option) unregister-option)
-      ((register-callback) register-callback)
-      ((unregister-callback-id) unregister-callback-id)
-      ((for-each) options-for-each)
-      ((for-each-general) options-for-each-general)
-      ((generate-restore-forms) generate-restore-forms)
-      ((scm->kvp) scm->kvp)
-      ((kvp->scm) kvp->scm)
-      ((touch) touch)
-      ((clear-changes) clear-changes)
-      ((run-callbacks) run-callbacks)
-      ((set-default-section) set-default-section)
-      ((get-default-section) get-default-section)
-      (else (gnc:warn "options: bad key: " key "\n"))))
-
-  dispatch)
-
-(define (gnc:register-option options new-option)
-  ((options 'register-option) new-option))
-
-(define (gnc:options-register-callback section name callback options)
-  ((options 'register-callback) section name callback))
-
-(define (gnc:options-register-c-callback section name c-callback data options)
-  (let ((callback (lambda () (gncp-option-invoke-callback c-callback data))))
-    ((options 'register-callback) section name callback)))
-
-(define (gnc:options-unregister-callback-id id options)
-  ((options 'unregister-callback-id) id))
-
-(define (gnc:options-for-each thunk options)
-  ((options 'for-each) thunk))
-
-(define (gnc:options-for-each-general section-thunk option-thunk options)
-  ((options 'for-each-general) section-thunk option-thunk))
-
-(define (gnc:lookup-option options section name)
-  (if options
-      ((options 'lookup) section name)
-      #f))
-
-(define (gnc:unregister-option options section name)
-  ((options 'unregister-option) section name))
-
-(define (gnc:generate-restore-forms options options-string)
-  ((options 'generate-restore-forms) options-string))
+(define-public (gnc:option-get-value book category key)
+  (define acc (if (pair? key) cons list))
+  (qof-book-get-option book (acc category key)))
 
-(define (gnc:options-scm->kvp options book clear-option?)
-  (if clear-option?
-      (qof-book-options-delete book '()))
-  ((options 'scm->kvp) book))
+(define-public (gnc:option-make-internal! options section name)
+  (let ((option (gnc-lookup-option (options 'lookup) section name)))
+    (and option (GncOption-make-internal option))))
 
-(define (gnc:options-kvp->scm options book)
-  ((options 'kvp->scm) book))
+(define-public (gnc:option-type option)
+  (GncOption-get-type option))
 
-(define (gnc:options-clear-changes options)
-  ((options 'clear-changes)))
+;; Used only by test-stress-options.scm
+(define-public (gnc:option-data option)
+  (let ((num-values (GncOption-num-permissible-values option))
+        (retval '()))
+    (do ((i 0 (1+ i))) ((>= i num-values))
+      (let ((value (GncOption-permissible-value option i))
+            (name (GncOption-permissible-value-name option i)))
+        (set! retval (cons retval (vector value name)))))
+    retval))
```

Comment:
Ok I forgot to reverse the result;

```scheme
(define-public (gnc:option-data option)
  (define num-values (GncOption-num-permissible-values option))
  (let lp ((i 0) (retval '()))
    (cond
     ((>= i num-values) retval)
     (else 
      (let ((value (GncOption-permissible-value option i))
            (name (GncOption-permissible-value-name option i)))
        (lp (1+ i) (cons (vector value name) retval))))))
```

or keeping it your style, you need to modify retval as such:

```scheme
(define-public (gnc:option-data option)
  (let ((num-values (GncOption-num-permissible-values option))
        (retval '()))
    (do ((i 0 (1+ i))) ((>= i num-values))
      (let ((value (GncOption-permissible-value option i))
            (name (GncOption-permissible-value-name option i)))
        (set! retval (append retval (list (vector value name))))))
    retval))
```

Reply:
@christopherlam How is
```scheme
(set! retval (append retval (list (vector value name))))))
```
Any different from 
```scheme
(set! retval (cons retval (vector value name)))))
```
besides verbosity?
Shouldn't it be
```scheme
     (set! retval (cons (vector value name) retval))))
(reverse retval))
```
to do the prepend & reverse?

Reply:
The `append` will take lists and return a list.

The `(set! retval (cons retval [x]))` will make `retval` a pair whose cdr is never the empty list `'()`, therefore retval is not a list. Try the following in a guile repl:
```scheme
(define retval '())
(set! retval (cons retval (vector 1 2)))
(set! retval (cons retval (vector 3 4)))
(set! retval (cons retval (vector 5 6)))
retval
$1 = (((() . #(1 2)) . #(3 4)) . #(5 6))
```
this ugly (() . x) form means retval is a nested pair tree whose cars are pairs and cdrs are vectors.

The `     (set! retval (cons (vector value name) retval)))) (reverse retval))` form is another acceptable form to prepend&reverse.

Reply:
Got it. For completeness
```scheme
scheme@(guile-user)> (define ls '())
scheme@(guile-user)> (set! ls (cons (vector 1 2) ls))
scheme@(guile-user)> (set! ls (cons (vector 3 4) ls))
scheme@(guile-user)> (set! ls (cons (vector 5 6) ls))
scheme@(guile-user)> ls
$1 = (#(5 6) #(3 4) #(1 2))
scheme@(guile-user)> (reverse ls)
$2 = (#(1 2) #(3 4) #(5 6))
```
showing that it does the right thing.

Reply:
> As for the let-loop implementation vs. set!, the set! is still there, it's just implicit.

Correct, however using `set!` in scheme has the disadvantage that it tells the compiler "this variable is mutated during execution" therefore it's stored into a box rather than some primitive variable. So, it adds an extra layer is indirection, and is not as optimisable during compilation. Guile (and racket etc) does some fancy `constexpr` type things during compilation nowadays.

Reply:
That doesn't really matter. All of the list nodes are going onto the heap regardless, so the only bit that's local and mutated is the pointer to the head of the stack. However once it's allocated its frequent use will keep it cached and it will be accessible just as fast as a stack variable is.

The [real reference](https://www.gnu.org/software/guile/manual/html_node/Variables-and-the-VM.html) about `set!` is a lot clearer about the effects of `set!`, though I find the claim that assigning to a variable means that the variable must be accessible from other closures unpersuasive; that's only true if the variable has references outside of the closure: From the example, `foo`, `a`, and `b` are various sorts of external references; if the lambda has a `let` variable `c` that gets `set!` called on it there's no possible external access to `c` and so no reason to box it.

But that's the decision Andy Wingo made, so that's what we have to live with. Considering that we're dealing with a linked list anyway where every node has to be on the heap allocating one more ptr for `retval` isn't a big deal and since it's touched every time through the loop it's going to stay in cache. He mentions indirection through a "box", but elswhere mentions indirections for other variables too so I'm not sure whether it makes a difference. I'm not going to dive the byte-interpreter code to find out, either.

Reply:
@christopherlam It seems that constructing lists either way is pretty expensive: With the tail recursion version test-stress-options takes 21.7 seconds on my mac (guile 2.2.7) and the `set!` version takes 23.5 seconds. The broken version takes only 5.5. 
```scheme
(define gnc:option-data option)
  (list (vector 1 2))
```
also takes only 5.5 seconds and still passes the test.
Interestingly in maint test-stress-options takes 11.8 seconds and if [test-stress-options.scm line 79-80](https://github.com/Gnucash/gnucash/blob/8a80993f57d615d4eab1d3b1eba8bbfeec73ca4c/gnucash/report/reports/standard/test/test-stress-options.scm#L79) is changed to `((multichoice) (list 1 2))` it takes only 3.5 seconds even though in maint `gnc:option-data` is a simple lookup.

Note that this test is the sole remaining call to `gnc:option-data`. It doesn't seem to me that it's worth the increase in unit test time.

Reply:
> Note that this test is the sole remaining call to gnc:option-data. It doesn't seem to me that it's worth the increase in unit test time.

Ok. Will there be a C++ API to retrieve the option values for multichoice?

Reply:
> 

PS of course changing option-data to return `(list 1 2)` will not make test-stress-options fail -- this test suite is designed to ensure that setting options will not cause a bad guile crash; setting multichoice to `1` or `2` will generally simply log a "One of the newer multichoice options '1' is not available, fallback to the option 'increasing'." warning and allow the report to run unimpeded.

Reply:
> Ok. Will there be a C++ API to retrieve the option values for multichoice?

Value(s) are retrieved with `option->get_value`. Permissible values on a multichoice, which is what `gnc:option-data` retrieved in that case, are available one at a time with `option->permissible_value(size_t index)` and the display name with `option->permissible_value_name(size_t index)`. Multichoice values are either a size_t or a std::vector<size_t> depending on whether the option is a multi-select.

> PS of course changing option-data to return (list 1 2) will not make test-stress-options fail -- this test suite is designed to ensure that setting options will not cause a bad guile crash; setting multichoice to 1 or 2 will generally simply log a "One of the newer multichoice options '1' is not available, fallback to the option 'increasing'." warning and allow the report to run unimpeded.

You're missing the point: The test in test-stress-options doesn't care what's in the list as long as it's not empty and its cdr is a pair. I suppose that might make sense when `gnc:option-data` retrieves the list-of-vectors containing the permissible values in the Scheme implementation, but there's not much point to that with a C++ implementation because the compiler enforces that permissible values is a std::vector<GncOptionMultichoiceChoice>. Attempting to insert something else in that vector will fail before the test can run.

Reply:
This may work to reduce dependency on `gnc:option-data`

```patch
diff --git a/gnucash/report/reports/standard/test/test-stress-options.scm b/gnucash/report/reports/standard/test/test-stress-options.scm
index 9a1951785..18c6d74d5 100644
--- a/gnucash/report/reports/standard/test/test-stress-options.scm
+++ b/gnucash/report/reports/standard/test/test-stress-options.scm
@@ -11,6 +11,7 @@
 (use-modules (gnucash reports))
 (use-modules (tests test-report-extras))
 (use-modules (srfi srfi-9))
+(use-modules (srfi srfi-26))
 (use-modules (srfi srfi-64))
 (use-modules (srfi srfi-98))
 (use-modules (tests srfi64-extras))
@@ -68,7 +69,7 @@
        (gnc:options-for-each
         (lambda (option)
           (when (case (gnc:option-type option)
-                  ((multichoice) (pair? (cdr (gnc:option-data option))))
+                  ((multichoice) (> (gnc-option-num-permissible-values option) 1))
                   ((boolean) #t)
                   (else #f))
             (set! report-options-tested
@@ -76,8 +77,9 @@
                      (gnc:option-section option)
                      (gnc:option-name option)
                      (case (gnc:option-type option)
-                       ((multichoice) (map (lambda (d) (vector-ref d 0))
-                                           (gnc:option-data option)))
+                       ((multichoice)
+                        (map (cut gnc-option-permissible-value option <>)
+                             (iota (gnc-option-num-permissible-values option))))
                        ((boolean) (list #t #f))))
                     report-options-tested))))
         options)
```

Reply:
Yes, by moving the implementation to its single point of use and so removing unneeded API.

> PS of course changing option-data to return (list 1 2) will not make test-stress-options fail -- this test suite is designed to ensure that setting options will not cause a bad guile crash; setting multichoice to 1 or 2 will generally simply log a "One of the newer multichoice options '1' is not available, fallback to the option 'increasing'." warning and allow the report to run unimpeded.

I misunderstood this--and the test suite--a month ago. The only way to make sense of the c++options extra time is that it's causing the reports to run twice as many times. Is the different way of creating the list of permissible values messing up the combination creation?

Reply:
> Yes, by moving the implementation to its single point of use and so removing unneeded API.
> 
> > PS of course changing option-data to return (list 1 2) will not make test-stress-options fail -- this test suite is designed to ensure that setting options will not cause a bad guile crash; setting multichoice to 1 or 2 will generally simply log a "One of the newer multichoice options '1' is not available, fallback to the option 'increasing'." warning and allow the report to run unimpeded.
> 
> I misunderstood this--and the test suite--a month ago. The only way to make sense of the c++options extra time is that it's causing the reports to run twice as many times. Is the different way of creating the list of permissible values messing up the combination creation?

Unlikely. The default test-stress-options does simple combinatorics i.e. if there are two options eg (1) open? true/false (2) colour? red/blue/green, the *simple* one will run 3 tests only (true red), (false blue), (true green).

Not sure why cpp would be slower. I'll see and benchmark this branch one of these days.